### ISSUE TITLE
Linear algebra sgemm_transAB crashes on unpredictable sizes

### ISSUE THREAD

**Issue by alexreinking** [halide/Halide#4631]
While porting the linear algebra examples to CMake I found that I get crashes with halide_benchmark, sgemm_transAB and sizes 288 and 544.

I haven't changed any actual code (save for some adjusted include paths), so I figured something must be wrong with the test.

While the version on master (via Makefile build) doesn't crash on _those_ two sizes, I have found that it crashes on many sizes: 193, 194, 197, 225-230, 232, 257-265, 267, 289-301, 321-333, 335, 353-368, 385-400, 402, 417-433, 435, 449-466, 468, 481-500, 513-533.

Over the same range, the version on my branch crashes on many more. 193-230, 232, 257-301, 321-368, 385-433, 435, 449-500, 513-544

**Comment by alexreinking** [comment]
I also get failures with halide_sgemv_notrans_32.

**Comment by abadams** [comment]
Having trouble reproing. What are you running exactly? The following seems to work for me:

```
make bin/halide_benchmarks && for ((i=100;i<1000;i++)); do ./bin/halide_benchmarks sgemm_transAB $i ; done
```

Also, can you try removing the no_asserts flag in the Makefile? That';s just there to shave a few cycles in the tiny-matrix case.

**Comment by alexreinking** [comment]
@abadams -- let me try again and get back to you. This was a few months ago.

**Comment by abadams** [comment]
Heh yeah. Quarantine means that Andrew slowly gets to the bottom of his TODO list.

**Comment by alexreinking** [comment]
@abadams - Have you tested with AVX-512? I thought I disabled it, but maybe I did something wrong. It's broken for me on Windows, WSL, and Linux, with both Make and CMake build systems

**Comment by alexreinking** [comment]
@abadams  - ok, yes, I screwed up. This is an AVX-512 issue.

**Comment by alexreinking** [comment]
I do still get errors on some sizes with AVX-512 disabled, but not as many. See the attached list of sizes on which sgemm_transAB fails for me.

[failing-avx512.txt](https://github.com/halide/Halide/files/4531497/failing-avx512.txt)
[failing-no-avx512.txt](https://github.com/halide/Halide/files/4531498/failing-no-avx512.txt)

**Comment by alexreinking** [comment]
Just tried with a fresh clone from master. Here's my (lightly edited) terminal output:

```
alex@alex-ubuntu:~/Development$ git clone https://github.com/halide/Halide.git hl-master
alex@alex-ubuntu:~/Development$ mkdir hl-master-build
alex@alex-ubuntu:~/Development$ cd hl-master-build
alex@alex-ubuntu:~/Development/hl-master-build$ echo $LLVM_CONFIG
/home/alex/Development/llvm/10.x/bin/llvm-config
alex@alex-ubuntu:~/Development/hl-master-build$ echo $CLANG
/home/alex/Development/llvm/10.x/bin/clang
alex@alex-ubuntu:~/Development/hl-master-build$ python3 -m venv venv
alex@alex-ubuntu:~/Development/hl-master-build$ . venv/bin/activate
(venv) alex@alex-ubuntu:~/Development/hl-master-build$ pip install -r ../hl-master/python_bindings/requirements.txt 
[...]
Installing collected packages: numpy, scipy, pillow, imageio, pybind11
Successfully installed imageio-2.8.0 numpy-1.18.3 pillow-7.1.1 pybind11-2.5.0 scipy-1.4.1
(venv) alex@alex-ubuntu:~/Development/hl-master-build$ $CC --version
gcc-9 (Ubuntu 9.2.1-17ubuntu1~18.04.1) 9.2.1 20191102
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

(venv) alex@alex-ubuntu:~/Development/hl-master-build$ $CXX --version
g++-9 (Ubuntu 9.2.1-17ubuntu1~18.04.1) 9.2.1 20191102
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

(venv) alex@alex-ubuntu:~/Development/hl-master-build$ make -f ../hl-master/Makefile -j 18
[...]
(venv) alex@alex-ubuntu:~/Development/hl-master-build$ make -f ../Halide/Makefile -j 18 build_apps
[...]
(venv) alex@alex-ubuntu:~/Development/hl-master-build$ cd bin/apps/linear_algebra
(venv) alex@alex-ubuntu:~/Development/hl-master-build/bin/apps/linear_algebra$ for ((i=100;i<1000;i++)); do ./bin/halide_benchmarks sgemm_transAB $i ; done
  Halide              sgemm_transAB     100           44.665924             23.0601
  [...]
  Halide              sgemm_transAB     192           61.680984             116.543
Segmentation fault (core dumped)
Segmentation fault (core dumped)
  Halide              sgemm_transAB     195          314.169266             23.9646
  Halide              sgemm_transAB     196          313.677234             24.3715
Segmentation fault (core dumped)
  Halide              sgemm_transAB     198          313.218406             25.1582
  [...]
  Halide              sgemm_transAB     224           93.218145             122.186
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     231          323.581172             38.5884
Segmentation fault (core dumped)
  Halide              sgemm_transAB     233          324.129422             39.5281
  [...]
  Halide              sgemm_transAB     256          120.139582             141.284
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     266          380.083781             50.0768
Segmentation fault (core dumped)
  Halide              sgemm_transAB     268          380.864906             51.1055
  [...]
  Halide              sgemm_transAB     288          161.513078             149.441
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     302          383.000875             72.6297
  [...]
  Halide              sgemm_transAB     320          169.606547             195.011
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     334          439.172328             85.6028
Segmentation fault (core dumped)
  Halide              sgemm_transAB     336          439.102359              87.159
  [...]
  Halide              sgemm_transAB     352          219.503000             200.389
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     369          462.333297             109.557
  [...]
  Halide              sgemm_transAB     384          227.819148             250.486
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     401          528.136766             123.005
Segmentation fault (core dumped)
  Halide              sgemm_transAB     403          537.854844             122.595
  [...]
  Halide              sgemm_transAB     416          300.679312             241.155
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     434          546.307047             150.669
Segmentation fault (core dumped)
  Halide              sgemm_transAB     436          544.007719             153.402
  [...]
  Halide              sgemm_transAB     448          315.002469             287.355
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     467          644.112156             159.137
Segmentation fault (core dumped)
  Halide              sgemm_transAB     469          632.349344             164.184
  [...]
  Halide              sgemm_transAB     480          411.020641             270.748
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     501          652.722094             193.811
  [...]
  Halide              sgemm_transAB     512          445.811844             302.828
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     534          595.646062              257.08
  [...]
  Halide              sgemm_transAB     544          539.373484              300.12
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     567          660.748375             277.335
  [...]
  Halide              sgemm_transAB     576          562.635031             341.426
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     599          753.931500             286.496
  [...]
  Halide              sgemm_transAB     608          756.075891             298.733
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     632          941.107750             269.506
  [...]
  Halide              sgemm_transAB     640          721.891719             364.837
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     664         1001.532156             293.628
  [...]
  Halide              sgemm_transAB     672          973.347531             313.166
Segmentation fault (core dumped)
  [...]
Segmentation fault (core dumped)
  Halide              sgemm_transAB     693         1783.986062             187.363
  [...]
  Halide              sgemm_transAB     999         4286.182500             233.307
```
